src/ccontrol/MergingHandler.cc

```diff
@@ -38,22 +43,269 @@
+                            throw runtime_error(context + "message size is 0 at offset " +
+                                                to_string(offset - sizeof(uint32_t)) + ", file: " + httpUrl);
+                        }
+                        if (msgSizeBytes > ProtoHeaderWrap::PROTOBUFFER_HARD_LIMIT) {
+                            throw runtime_error(context + "message size " + to_string(msgSizeBytes) +
+                                                " at offset " + to_string(offset - sizeof(uint32_t)) +
+                                                " exceeds the hard limit of " +
+                                                to_string(ProtoHeaderWrap::PROTOBUFFER_HARD_LIMIT) +
+                                                ", file: " + httpUrl);
+                        }
+                        // Extend the message buffer (if needed). Note that buffer never gets
+                        // truncated to avoid excessive memory deallocations/allocations.
+                        if (msgBufSize < msgSizeBytes) {
+                            msgBufSize = msgSizeBytes;
+                            msgBuf.reset(new char[msgBufSize]);
+                        }
+                    }
+                } else {
+                    // Continue or finish reading the message body.
+                    size_t const bytes2read =
+                            std::min((size_t)msgSizeBytes - msgBufNext, (size_t)(end - next));
+                    std::memcpy(msgBuf.get() + msgBufNext, next, bytes2read);
+                    next += bytes2read;
+                    offset += bytes2read;
+                    msgBufNext += bytes2read;
+                    if (msgBufNext == msgSizeBytes) {
+                        // Done reading message body.
+                        msgBufNext = 0;
+                        // Parse and evaluate the message.
+                        bool const success = messageIsReady(msgBuf.get(), msgSizeBytes);
+                        if (!success) {
+                            throw runtime_error(context + "message processing failed at offset " +
+                                                to_string(offset - msgSizeBytes) + ", file: " + httpUrl);
+                        }
+                        // Reset the variable to prepare for reading the next header & message (if any).
+                        msgSizeBytes = 0;
+                    }
+                }
+            }
+        });
+        if (msgSizeBufNext != 0) {
+            throw runtime_error(context + "short read of the message header at offset " +
+                                to_string(offset - msgSizeBytes) + ", file: " + httpUrl);
+        }
+        if (msgBufNext != 0) {
+            throw runtime_error(context + "short read of the message body at offset " +
+                                to_string(offset - msgSizeBytes) + ", file: " + httpUrl);
+        }
+    } catch (exception const& ex) {
```

Comment:
Unfortunately, it would be hard to do because the class `HttpReader` that's employed by the algorithm throws standard exceptions, including `std::invalid_argument`, and `std::runtime_error`.  Besides, regardless of the origin and type of exception, I would like this method not to throw anything to adhere to the error processing logic of the callee (the method `MergingHandlker::flush`) which expects the algorithm to return `false` in case of any failures. The third reason for catching all exceptions is to ensure the garbage collection step that's fired right after that block is guaranteed to be executed:
```
// Remove the file from the worker if it still exists. Report and ignore errors.
    // The files will be garbage-collected by workers.
    try {
        HttpClient remover("DELETE", httpUrl);
        remover.read([](char const* inBuf, size_t inBufSize) {});
    } catch (exception const& ex) {
        LOGS(_log, LOG_LVL_WARN, context << "failed to remove " << httpUrl << ", ex: " << ex.what());
    }
    return success;
```